### PR TITLE
publish MapState, MaplibreMap(state), and MaplibreRuntime

### LINKED ISSUE #18: Support direct mutation of base style
Based on a request in Slack; there's some use cases for the consumer to want to change properties of a base style layer at runtime. Currently this is "supported" only by using `Anchor.Replace` to replace that layer and mimic its properties. Could we in addition, provide some mutable state control surface for a map style? And if so, how do we resolve conflicts with the current style composition for adding layers?

### LINKED ISSUE #28: Support snapshots
* iOS: https://maplibre.org/maplibre-native/ios/latest/documentation/maplibre-native-for-ios/staticsnapshotexample/
* Android: https://maplibre.org/maplibre-native/android/examples/snapshotter/

We should provide snapshot functionality to a similar extent as the native SDKs.

This may also come in handy for automated testing.

### LINKED ISSUE #374: Consider moving the map style parameters to the StyleState
The map snapshotter (#266) introduces an additional use case for map style. Conceptually, it makes sense that:
1. a snapshotter should be able to have the same style as an interactive map
2. a snapshotter's style should be configurable the same as an interactive map

So, perhaps it would make sense for the StyleState, instead of the MapLibreMap, to own the base style URI and the style composition, and the StyleState to be passed as parameter to the snapshotter?

### PR DESCRIPTION
**Pull request by sargunv**
## Description

Steps 4 and 6 of `.agents/docs/API_REDESIGN.md` plus the final quality sweep: the public `MapState` API (camera, style content, sources/layers ownership collections, `styleErrors`, `isAttached`, detached `snapshot`), `rememberMapState`/`MaplibreMap(state)`/`LocalMapState`, `MaplibreRuntime`, and full migration of the demo app, overlays, material3, and docs — then a whole-stack simplification and test-quality pass driven by two structural audits.

## Validation

All Linux gates green per commit; four Codex review rounds converged, including a sanity pass over the sweep whose six regressions (five in the new snapshot path) are fixed with pinning tests. Bonus bug fixes surfaced by the work: painter-SDF images uploaded unconverted since #720, and `toImageBitmap` channel order was wrong on non-Android platforms.

## AI assistance

Claude Code with Fable 5 planning and Fable 5 / Opus 5 subagents implementing; Codex (GPT) as external reviewer.## Issues

Closes #28 — `MapState.snapshot` renders a still image with no composable attached on the native platforms; web declines, matching GL JS.
Closes #18 — `MapState.layers[id]` writes reach the live style; composition-owned ids reject writes, answering the conflict question by ownership.
Closes #374 — the base style and style content live on `MapState`, which is what a snapshotter consumes.
Unblocks #448 — the offline API lives on the public application-scoped `MaplibreRuntime`, the owner a database merge needed.
Related: #538 — concrete progress: imperative camera, projections, feature queries, and live layer and source access cover most of what callers reached past the wrapper for; the general escape hatch lands in the next stacked PR.

### CODE REVIEW

File: lib/maplibre-compose/src/maplibreNativeMain/kotlin/org/maplibre/compose/map/MlnFfiMapEngine.kt [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+  ): ImageBitmap {
+    ensureMlnFfiConfigured()
+    state.ensureBaseStyleSelected()
+    val target = createSnapshotTarget()
+    try {
+      // A live core keeps its loaded style and scale; a detached bare state gets a fresh one,
+      // allocated under the lock as the reservation holder so a close cannot orphan it.
+      val core = sessionLock.withLock {
+        core
+          ?: acquireCoreLocked(
```
[the remark this answers is not in the document]
Comment (sargunv) [reply lib/maplibre-compose/src/maplibreNativeMain/kotlin/org/maplibre/compose/map/MlnFfiMapEngine.kt:152]: Fixed on the stack tip (#1120), commit 2f5ae8cf: the snapshot always acquires through the eviction-safe path, recreating a retained core whose backend differs from the snapshot target; pinned by MlnFfiRetainedCoreTest.

File: lib/maplibre-compose/src/maplibreNativeMain/kotlin/org/maplibre/compose/map/MlnFfiMapEngine.kt [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+      target.close()
+    }
+  }
+
+  private suspend fun awaitStyleLoaded(
+    core: MlnFfiMapCore,
+    deadline: TimeSource.Monotonic.ValueTimeMark,
+    timeout: Duration,
+  ) {
+    while (!core.hasLoadedFirstStyle) {
```
[the remark this answers is not in the document]
Comment (sargunv) [reply lib/maplibre-compose/src/maplibreNativeMain/kotlin/org/maplibre/compose/map/MlnFfiMapEngine.kt:186]: Fixed on the stack tip (#1120), commit 2f5ae8cf: the wait rides a style generation (bumped on selection, published on load) instead of the sticky first-load flag; pinned by MlnFfiStyleGenerationTest.

File: lib/maplibre-compose/src/commonMain/kotlin/org/maplibre/compose/map/MapState.kt [hunk 3]
```diff
@@ -142,129 +215,362 @@ internal class MapState(
...
+    }
+    sources.refreshSources()
   }
 
   /** Unwires the session; the state, its content, and its desired style survive for the next. */
   internal fun detachSession() {
-    cameraState.map = null
-    updateBinding(null)
-    styleState.detach()
+    adapterState.value = null
```
[the remark this answers is not in the document]
Comment (sargunv) [reply lib/maplibre-compose/src/commonMain/kotlin/org/maplibre/compose/map/MapState.kt:506]: Fixed on the stack tip (#1120), commit 2f5ae8cf: detach resets the per-composition hooks to inert defaults, so a retained core's later load event fires nothing; the next attach rewires.

File: lib/maplibre-compose/src/maplibreNativeMain/kotlin/org/maplibre/compose/map/MlnFfiMapEngine.kt [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+ * because that is the first moment the density and the render backend are known, and it survives
+ * detach so a re-entering composition re-attaches to the live map instead of recreating it.
+ */
+internal actual class MapEngine actual constructor(private val state: MapState) : AutoCloseable {
+
+  /**
+   * The core keeps its loaded style across detach, so the state keeps its binding while it lives.
+   */
+  actual val detachedAdapter: MapAdapter?
+    get() = core
```
[the remark this answers is not in the document]
Comment (sargunv) [reply lib/maplibre-compose/src/maplibreNativeMain/kotlin/org/maplibre/compose/map/MlnFfiMapEngine.kt:32]: Fixed on the stack tip (#1120), commit 2f5ae8cf: releasing a session (and closing a snapshot target) resets the attached-viewport flag, so a re-attach at new dimensions never trusts stale extent.

File: demo-app/common/src/maplibreNativeMain/kotlin/org/maplibre/compose/docsnippets/Offline.kt [hunk 5]
```diff
@@ -5,18 +5,21 @@ package org.maplibre.compose.docsnippets
...
 import org.maplibre.compose.offline.OfflinePackDefinition
-import org.maplibre.compose.offline.rememberOfflineManager
+import org.maplibre.compose.runtime.MaplibreRuntime
 import org.maplibre.spatialk.geojson.BoundingBox
 
 @Composable
 fun Offline() {
   // #region manager
-  val offlineManager = rememberOfflineManager()
+  val offlineManager = remember { MaplibreRuntime.offline }
```
[the remark this answers is not in the document]
Comment (sargunv) [reply demo-app/common/src/maplibreNativeMain/kotlin/org/maplibre/compose/docsnippets/Offline.kt]: Outdated as marked: the snippet was rewritten during the runtime-object migration; on Android an unconfigured MaplibreRuntime.offline now fails with a message naming MapLibre.configure(context) rather than crashing opaquely.

File: lib/maplibre-compose/src/maplibreNativeMain/kotlin/org/maplibre/compose/map/MlnFfiMapEngine.kt [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+            scaleFactor = state.density.density.toDouble(),
+            layoutDirection = state.layoutDirection,
+            backend = target.backend,
+          )
+      }
+      // A session attach pushes the selected style; a snapshot has no session, so it pushes it
+      // here. The core drops a style it already has.
+      core.setBaseStyle(state.baseStyle)
+      core.start()
+      core.setCameraPosition(state.camera)
```
[the remark this answers is not in the document]
Comment (sargunv) [reply lib/maplibre-compose/src/maplibreNativeMain/kotlin/org/maplibre/compose/map/MlnFfiMapEngine.kt:162]: Fixed on the stack tip (#1120), commit 2f5ae8cf: the snapshot path neutralizes retained session constraints (bounds, zoom and pitch ranges, padding) before applying the state's camera; a later attach restores them through applyOptions.

File: lib/maplibre-compose/src/commonMain/kotlin/org/maplibre/compose/map/MapState.kt [hunk 7]
```diff
@@ -142,129 +211,336 @@ internal class MapState(
...
+  private fun Expression<BooleanValue>.compileOrNull(): CompiledExpression<BooleanValue>? =
+    takeUnless {
+      it == const(true)
+    }
+    ?.compile(ExpressionContext.None)
+
+  /** The per-composable session options; attach applies them, and a change reaches a live map. */
+  internal var sessionOptions: SessionOptions? = null
+    set(value) {
+      if (value == field) return
```
[the remark this answers is not in the document]
Comment (sargunv) [reply lib/maplibre-compose/src/commonMain/kotlin/org/maplibre/compose/map/MapState.kt:481]: Fixed on the stack tip (#1120), commit 2f5ae8cf: camera padding resolves against the layout direction at capture, so a direction flip changes the options and reapplies directional padding.

File: lib/maplibre-compose/src/commonMain/kotlin/org/maplibre/compose/map/MapState.kt [hunk 8]
```diff
@@ -142,129 +211,336 @@ internal class MapState(
...
   internal fun detachSession() {
-    cameraState.map = null
-    updateBinding(null)
-    styleState.detach()
+    adapterState.value = null
+    // a snapshot kept past detachment would report a viewport no map is showing
+    viewportState.value = null
+    // An engine that keeps the map alive keeps its loaded binding and the applied snapshot too.
+    if (engine.detachedAdapter == null) updateBinding(null)
+    sources.clear()
```
[the remark this answers is not in the document]
Comment (sargunv) [reply lib/maplibre-compose/src/commonMain/kotlin/org/maplibre/compose/map/MapState.kt:511]: Fixed on the stack tip (#1120), commit 2f5ae8cf: source collections now clear only when the binding actually unloads, matching layers; a native detach with a retained core keeps them.

File: lib/maplibre-compose/src/maplibreNativeMain/kotlin/org/maplibre/compose/map/MlnFfiMapEngine.kt [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+      awaitStyleLoaded(core, deadline, timeout)
+      // One sync of the desired style content against the loaded style before rendering, so
+      // content set on a detached state reaches the image.
+      state.host.requestApplyChanges()
+      state.host.awaitPendingWork()
```
[the remark this answers is not in the document]
Comment (sargunv) [reply lib/maplibre-compose/src/maplibreNativeMain/kotlin/org/maplibre/compose/map/MlnFfiMapEngine.kt:167]: Fixed on the stack tip (#1120), commit 5ffbc923: the snapshot publishes its target's viewport to the state and runs one more content sync before the sealed frame; pinned by a pixel test composing a layer only when the viewport exists.

File: lib/maplibre-compose/src/maplibreNativeMain/kotlin/org/maplibre/compose/map/MlnFfiMapView.kt [hunk 10]
```diff
@@ -80,56 +68,43 @@ internal fun MlnFfiMapView(
...
+    sessionState.attachSession(core)
     core.start()
   }
 
   DisposableEffect(session) {
     onDispose {
       session.close()
-      currentOnReset.value()
+      engine.releaseSession(session)
+      sessionState.detachSession()
```
[the remark this answers is not in the document]
Comment (sargunv) [reply lib/maplibre-compose/src/maplibreNativeMain/kotlin/org/maplibre/compose/map/MlnFfiMapView.kt:107]: Fixed on the stack tip (#1120), commit 5ffbc923: a detached map reports isCameraMoving false and pauses the animation; the suspended call survives and completes on re-attach, and close ends it promptly — both pinned by tests.
